fedimint-core/src/module/audit.rs

```diff
@@ -59,11 +63,156 @@ impl Display for Audit {
 pub struct AuditItem {
     pub name: String,
     pub milli_sat: i64,
+    pub module_name: String,
 }
 
 impl Display for AuditItem {
     fn fmt(&self, formatter: &mut Formatter) -> std::fmt::Result {
         let sats = (self.milli_sat as f64) / 1000.0;
-        formatter.write_fmt(format_args!("{:>+15.3}|{}", sats, self.name))
+        formatter.write_fmt(format_args!(
+            "{:>+15.3}|{:>10}|{}",
+            sats, self.module_name, self.name
+        ))
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct AuditSummary {
+    pub total_assets: i64,
+    pub total_liabilities: i64,
+    pub net_assets: i64,
+    pub module_summaries: HashMap<String, ModuleSummary>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct ModuleSummary {
+    pub assets: i64,
+    pub liabilities: i64,
+    pub net_assets: i64,
+}
+
+impl AuditSummary {
+    pub fn from_audit(audit: &Audit) -> Self {
+        let (total_assets, total_liabilities) =
+            sum_assets_and_liabilities(audit.items.iter().collect());
+        AuditSummary {
+            total_assets,
+            total_liabilities,
+            // liabilities are negative, so we add to calculate net assets
+            net_assets: total_assets + total_liabilities,
+            module_summaries: generate_module_summaries(&audit.items),
+        }
     }
 }
+
+fn generate_module_summaries(audit_items: &[AuditItem]) -> HashMap<String, ModuleSummary> {
+    let mut audit_items_by_module: HashMap<String, Vec<&AuditItem>> = HashMap::new();
+    audit_items.iter().for_each(|item| {
+        audit_items_by_module
+            .entry(item.module_name.clone())
+            .or_insert(Vec::new())
+            .push(item)
+    });
```

Comment:
Ahh that's perfect, thanks!

I originally tried using [`group_by`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.group_by), but that seems like a footgun since it groups by _consecutive_ elements with the same key, which makes calls to `add_items` fragile.